graphql/src/execution/execution.rs

```diff
@@ -57,6 +58,100 @@ impl CacheByBlock {
     }
 }
 
+struct WeightedResult {
+    result: Arc<QueryResult>,
+    weight: usize,
+}
+
+impl CacheWeight for WeightedResult {
+    fn indirect_weight(&self) -> usize {
+        self.weight
+    }
+}
+
+impl Default for WeightedResult {
+    fn default() -> Self {
+        WeightedResult {
+            result: Arc::new(QueryResult::new(Some(q::Value::Null))),
+            weight: 0,
+        }
+    }
+}
+
+struct QueryBlockCache(Vec<(String, VecDeque<CacheByBlock>)>);
+
+impl QueryBlockCache {
+    fn insert(
+        &mut self,
+        network: &str,
+        block_ptr: EthereumBlockPointer,
+        key: QueryHash,
+        result: Arc<QueryResult>,
+        weight: usize,
+    ) -> bool {
+        // Get or insert the cache for this network.
+        let cache = match self
+            .0
+            .iter_mut()
+            .find(|(n, _)| n == network)
+            .map(|(_, c)| c)
+        {
+            Some(c) => c,
+            None => {
+                self.0.push((network.to_owned(), VecDeque::new()));
+                &mut self.0.last_mut().unwrap().1
+            }
+        };
+
+        // If there is already a cache by the block of this query, just add it there.
+        let mut cache_insert = false;
+        if let Some(cache_by_block) = cache.iter_mut().find(|c| c.block == block_ptr) {
+            cache_insert = cache_by_block.insert(key, result.cheap_clone(), weight);
+        } else if *QUERY_CACHE_BLOCKS > 0 {
+            // We're creating a new `CacheByBlock` if:
+            // - There are none yet, this is the first query being cached, or
+            // - `block_ptr` is of higher or equal number than the most recent block in the cache.
+            // Otherwise this is a historical query which will not be cached.
+            let should_insert = match cache.iter().next() {
+                None => true,
+                Some(highest) if highest.block.number <= block_ptr.number => true,
```

Comment:
This can be simplified a bit:

```
let should_insert = match cache.iter().next() {
    None => true,
    Some(highest) => highest.block.number <= block_ptr.number,
}
```
Removes 1 branch from the `match` expression

Reply:
That was in the existing code that I just copied around, but I agree - I just pushed a change to address that.